### PR TITLE
lorawan: add channels mask configuration

### PR DESCRIPTION
**Pull request by JefersonFernando**
This PR add channels mask configuration for LoRaWAN, creating a request for LoRaMAC. This change is needed because using  LoRaWAN with ABP activation without configuring channels mask will provide no acceptable channels on data transmit, causing data loss.
Using TheThingsNetwork I notice that the device was losing a lot of Uplinks messages. Further investigation pointed that the channels masks used by my region (AU915 FSB 2) was using only the second frequency band. The lorawan subsys did not had a channels mask configuration and was using all available channels. A mismatch of this configuration implies wrong channels being selected on Uplinks and the gateway discards it.

### CODE REVIEW

File: subsys/lorawan/Kconfig [hunk 1]
```diff
@@ -44,35 +44,57 @@ config LORAWAN_PUBLIC_NETWORK
 	  Enable this option to use a public LoRaWAN network.
 	  Disable for private LoRaWAN networks.
 
-config LORAMAC_REGION_AS923
-	bool "Asia 923MHz Frequency band"
+choice
```
Comment (JordanYates) [inline subsys/lorawan/Kconfig:47]: The region configuration should be returned to how it was

File: subsys/lorawan/Kconfig [hunk 2]
```diff
@@ -44,35 +44,57 @@ config LORAWAN_PUBLIC_NETWORK
...
+		bool "North America 915MHz Frequency band"
 
-config LORAMAC_REGION_RU864
-	bool "Russia 864MHz Frequency band"
+	config LORAMAC_REGION_RU864
+		bool "Russia 864MHz Frequency band"
+
+endchoice
+
+config LORAWAN_CHANNELS_MASK_SIZE
```
Comment (JordanYates) [inline subsys/lorawan/Kconfig:82]: This should be a enum in the lorawan header, not a kconfig value.
Multiple regions can be enabled, which this does not express well.

File: subsys/lorawan/lorawan.c [hunk 3]
```diff
@@ -472,6 +472,26 @@ int lorawan_set_class(enum lorawan_class dev_class)
 	return 0;
 }
 
+int lorawan_set_channels_mask(uint16_t *channels_mask, uint8_t channels_mask_size)
+{
+	MibRequestConfirm_t mib_req;
+
+	if ((channels_mask == NULL) || (channels_mask_size != CONFIG_LORAWAN_CHANNELS_MASK_SIZE)) {
```
Comment (JordanYates) [inline subsys/lorawan/lorawan.c:479]: If length validation is still desired, it should be done depending on the currently enabled region, not a hardcoded value.

File: include/zephyr/lorawan/lorawan.h [hunk 4]
```diff
@@ -295,6 +295,16 @@ int lorawan_set_conf_msg_tries(uint8_t tries);
  */
 void lorawan_enable_adr(bool enable);
 
+/**
+ * @brief Set the channels mask.
+ *
```
Comment (JordanYates) [inline include/zephyr/lorawan/lorawan.h:300]: It would be good to add some notes here about when you would want to use this function

**Review by JordanYates — CHANGES_REQUESTED** [review]

File: subsys/lorawan/Kconfig (hunk 1 above)
Comment (JefersonFernando) [reply subsys/lorawan/Kconfig]: Done.

File: include/zephyr/lorawan/lorawan.h (hunk 4 above)
Comment (JefersonFernando) [reply include/zephyr/lorawan/lorawan.h:316]: Fixed.

File: subsys/lorawan/lorawan.c (hunk 3 above)
Comment (JefersonFernando) [reply subsys/lorawan/lorawan.c]: Fixed.

File: subsys/lorawan/Kconfig (hunk 2 above)
Comment (JefersonFernando) [reply subsys/lorawan/Kconfig]: Fixed.

**Review by JordanYates — CHANGES_REQUESTED** [review]
CI failures need to be fixed

**Review by JordanYates — DISMISSED** [review]

File: include/zephyr/lorawan/lorawan.h [hunk 5]
```diff
@@ -295,6 +311,20 @@ int lorawan_set_conf_msg_tries(uint8_t tries);
...
+ * Change the default channels mask. When mask is not changed, all the channels
+ * can be used for data transmition. Some Network Servers don't use all channels,
+ * in this case, the channels mask must be provided.
+ *
+ * @param channels_mask Buffer with channels mask to be used.
+ * @param channels_mask_size Size of channels mask buffer.
+ *
+ * @return 0 if successful, negative errno code if failure
```
Comment (rodrigopex) [inline include/zephyr/lorawan/lorawan.h:324]: Add the @ret val for each return code. 
```suggestion
 * @retval 0 successful
 * @retval -EXXXX <justify>
```

File: subsys/lorawan/lorawan.c [hunk 6]
```diff
@@ -472,6 +495,26 @@ int lorawan_set_class(enum lorawan_class dev_class)
 	return 0;
 }
 
+int lorawan_set_channels_mask(uint16_t *channels_mask, uint8_t channels_mask_size)
```
Comment (rodrigopex) [inline subsys/lorawan/lorawan.c:498]: Please add some tests to cover this function. The current test suite seems to have that untouched.

File: include/zephyr/lorawan/lorawan.h [hunk 7]
```diff
@@ -295,6 +311,20 @@ int lorawan_set_conf_msg_tries(uint8_t tries);
  */
 void lorawan_enable_adr(bool enable);
 
+/**
+ * @brief Set the channels mask.
+ *
+ * Change the default channels mask. When mask is not changed, all the channels
+ * can be used for data transmition. Some Network Servers don't use all channels,
```
Comment (rodrigopex) [inline include/zephyr/lorawan/lorawan.h:318]:
```suggestion
 * can be used for data transmission. Some Network Servers don't use all the channels,
```

File: include/zephyr/lorawan/lorawan.h [hunk 8]
```diff
@@ -295,6 +311,20 @@ int lorawan_set_conf_msg_tries(uint8_t tries);
...
+ * in this case, the channels mask must be provided.
+ *
+ * @param channels_mask Buffer with channels mask to be used.
+ * @param channels_mask_size Size of channels mask buffer.
+ *
+ * @return 0 if successful, negative errno code if failure
+ */
+int lorawan_set_channels_mask(uint16_t *channels_mask, uint8_t channels_mask_size);
```
Comment (rodrigopex) [inline include/zephyr/lorawan/lorawan.h:326]:
```suggestion
int lorawan_set_channels_mask(uint16_t *channels_mask, size_t channels_mask_size);
```

**Review by rodrigopex — CHANGES_REQUESTED** [review]

File: include/zephyr/lorawan/lorawan.h (hunk 5 above)
Comment (JefersonFernando) [reply include/zephyr/lorawan/lorawan.h]: Fixed.

File: subsys/lorawan/lorawan.c (hunk 6 above)
Comment (JefersonFernando) [reply subsys/lorawan/lorawan.c]: Fixed.

File: include/zephyr/lorawan/lorawan.h (hunk 7 above)
Comment (JefersonFernando) [reply include/zephyr/lorawan/lorawan.h]: Fixed.

File: include/zephyr/lorawan/lorawan.h (hunk 8 above)
Comment (JefersonFernando) [reply include/zephyr/lorawan/lorawan.h]: Fixed.

**Review by JefersonFernando — COMMENTED** [review]
Fixed channels mask set function returns, comments and add tests.

**Review by rodrigopex — APPROVED** [review]

**Review by JordanYates — APPROVED** [review]

**Pull request merged**